lister.cc

```diff
@@ -75,3 +77,50 @@ future<> lister::rmdir(fs::path dir) {
         return remove_file(dir.native());
     });
 }
+
+directory_lister::~directory_lister() {
+    if (_lister) {
+        on_internal_error(llogger, "directory_lister not closed when destroyed");
+    }
+}
+
+future<std::optional<directory_entry>> directory_lister::get() {
+    if (!_done) {
+        auto f = co_await open_checked_directory(general_disk_error_handler, _dir.native());
+        auto walker = [this] (fs::path dir, directory_entry de) {
+            return _queue.push_eventually(std::make_optional<directory_entry>(std::move(de)));
+        };
+        assert(!_lister);
+        _lister = std::make_unique<lister>(std::move(f), _type, std::move(walker), _filter, _dir, _do_show_hidden);
+        _done = _lister->done().then_wrapped([this] (future<> f) {
+            if (f.failed()) [[unlikely]] {
+                _queue.abort(f.get_exception());
+                return make_ready_future<>();
+            }
+            return _queue.push_eventually(std::nullopt);
+        }).finally([this] {
+            _lister.reset();
+        });
+    }
+    auto ret = co_await _queue.pop_eventually();
+    if (!ret) {
+        // In case the caller tries to get() again
+        // after returning end of list.
```

Comment:
Aborting the queue after it's exhausted will leave an exception for the *next* call to `get()`.
In that case, if `get()` is called yet-again (and I should probably add a test case for that),
`pop_eventually` should return an error since the queue is aborted.  (yes, we/I have made aborting the seastar::queue stateful so first thing `pop_eventually` does is to check if there's an exception set on the queue)

Reply:
Improved documentation in v4.

Reply:
And added a unit test for calling get() after the lister is done.

Reply:
I see now what I missed - you used a queue of optionals - so at the real end you pop a disengaged optional - here you don't pop anything (pop itself returns nothing) which means the disengaged optional was already popped last time. So ok.